PXB-2797: Schema mismatch when importing table with full-text index f…

aybek opened the pull request:
…rom xtrabackup backup

https://jira.percona.com/browse/PXB-2797

When importing a single table (IMPORT TABLESPACE) from a backup made using xtrabackup and the table contains a full-text index the import process will error out with: `ERROR 1808 (HY000) at line 132: Schema mismatch (Index xxxxxx field xxxxxx is ascending which does not match metadata file which is descending)`

The problem occurs because FTS tables are created internally by InnoDB Storage engine and they do not choose any ordering flag. Default is UNDEF. This patch solves the issue by treating UNDEF order as ASC when creating dict_table_t on `--prepare` step.

it-percona-cla commented:
[![CLA assistant check](https://cla-assistant.percona.com:443/pull/badge/signed)](https://cla-assistant.percona.com:443/percona/percona-xtrabackup?pullRequest=1526) <br/>All committers have signed the CLA.

altmannmarcelo commented on storage/innobase/dict/dict0dd.cc:
```diff
@@ -528,7 +528,7 @@ dict_table_t *dd_table_create_on_dd_obj(const dd::Table *dd_table,
     if (dd_col->is_virtual()) {
       n_v_cols++;
       if (dd_col->is_array()) {
-         n_m_v_cols++;
```
This might be your editor, which is adjusting the whole file and looking for formatting issues. Please disable it.

aybek replied:
yes, I think it is clang formatter doing this automatically. I configured it for PS should not we use it for PXB ?

altmannmarcelo commented on storage/innobase/dict/dict0dd.cc:
```diff
@@ -898,8 +898,7 @@ dict_table_t *dd_table_create_on_dd_obj(const dd::Table *dd_table,
   if (table->has_instant_drop_cols()) {
-    fill_dict_dropped_columns(&dd_table->table(), table,
-                              IF_DEBUG(crv, ) heap);
+    fill_dict_dropped_columns(&dd_table->table(), table, IF_DEBUG(crv, ) heap);
```
same here

altmannmarcelo commented on storage/innobase/dict/dict0dd.cc:
```diff
@@ -980,8 +979,11 @@ dict_table_t *dd_table_create_on_dd_obj(const dd::Table *dd_table,
         if (c == dd_col) break;
         if (c->is_virtual() == dd_col->is_virtual()) col_pos++;
       }
+      /* FULLTEXT and HASH indexes can have UNDEF order, we should treat UNDEF
```
anything PXB changes on areas that comes from upstream, we should use pxb macros:

either:
``` 
#ifdef XTRABACKUP
      bool is_asc = (idx_elem->order() == dd::Index_element::ORDER_ASC ||
                     idx_elem->order() == dd::Index_element::ORDER_UNDEF);
#else
bool is_asc = (idx_elem->order() == dd::Index_element::ORDER_ASC);
#endif  //XTRABACKUP
```

or

```
bool is_asc = (idx_elem->order() == dd::Index_element::ORDER_ASC IF_XB( || idx_elem->order() == dd::Index_element::ORDER_UNDEF));
```

satya-bodapati replied:
this entire function is PXB specific.

satya-bodapati replied:
so this ifdef might not be necessary

altmannmarcelo replied:
seems like code from before the macro rule.

satya-bodapati replied:
right.. since we are touching this, we can put ifdef around the entire function

aybek replied:
ack, will put it around the function

altmannmarcelo commented on storage/innobase/dict/dict0dd.cc:
```diff
@@ -2388,9 +2390,9 @@ static bool format_validate(THD *thd, const TABLE *form, row_type real_type,
-                                        : srv_page_size <= UNIV_ZIP_SIZE_MAX
-                                              ? "innodb_file_per_table=OFF"
-                                              : "innodb_page_size>16k";
+                            : srv_page_size <= UNIV_ZIP_SIZE_MAX
```
same here with regards to formatting

altmannmarcelo commented on storage/innobase/xtrabackup/test/t/pxb-2797.sh:
```diff
@@ -0,0 +1,34 @@
+# PXB-2797: Schema mismatch when importing table with full-text index from backup
+#
+
+start_server
```
Can you extend the test case and make sure other types of index that can have ASC or DESC are working properly?

aybek replied:
ack

aybek replied:
the only other index type that can have DESC is Hash index, but it is not supported by PXB storage engines. So I think no need to test this case

altmannmarcelo replied:
One nit picking I missed from the last review. Can we rename the test case to something more meaningful? We are trying to avoid just plain JIRA issue as the name. Something like pxb-2797-export_fulltext.sh will do it.

altmannmarcelo reviewed (changes requested):
@aybek - Congrats on your first PR :tada: 

A few comments inline in the files.
Also:
1. This fix comes from a shared code between the server and pxb. I assume the same issue will happen in the server? If so, we should raise it to upstream and submit a patch, otherwise, can you explain why server does not cross this code path?
2. Please submit a jenkins. Reach out to me over slack and I can explain how to do it.

satya-bodapati commented on storage/innobase/xtrabackup/test/t/pxb-2797.sh:
```diff
@@ -0,0 +1,34 @@
+
+start_server
+
+mysql -e "SET FOREIGN_KEY_CHECKS=0;"
```
no FKs on table, so this is not necessary

aybek replied:
ack

aybek commented:
Jenkins run - https://pxb.cd.percona.com/view/PXB%208.0/job/percona-xtrabackup-8.0-multijob/324/

altmannmarcelo reviewed (changes requested):
One minor change and the rest is good.

altmannmarcelo reviewed (approved):
LGTM. Thanks for working on it @aybek

satya-bodapati commented on storage/innobase/xtrabackup/test/t/pxb-2797_import_fulltext_schema_mismatch.sh:
```diff
@@ -0,0 +1,32 @@
+
+start_server
+
+mysql test <<EOF
+CREATE TABLE test.t1 (
+  t1_id int NOT NULL AUTO_INCREMENT,
+  t1_text varchar(50) DEFAULT NULL,
+  PRIMARY KEY (t1_id),
+  FULLTEXT KEY t1_textx (t1_text)
+) ENGINE=InnoDB;
+INSERT INTO test.t1 (t1_text) values ("text1");
```
testname need not have bug number.
You can add this under import.sh

aybek replied:
can't find import.sh file in xtrabackup tests
do you mean rename the file to import.sh ?

satya-bodapati replied:
Please use t/xb_export.sh

satya-bodapati commented on storage/innobase/xtrabackup/test/t/pxb-2797_import_fulltext_schema_mismatch.sh:
```diff
@@ -0,0 +1,32 @@
+  t1_text varchar(50) DEFAULT NULL,
+  PRIMARY KEY (t1_id),
+  FULLTEXT KEY t1_textx (t1_text)
+) ENGINE=InnoDB;
```
also marce asked for tables with explicit ASC and DESC columns

aybek replied:
done

aybek commented:
https://pxb.cd.percona.com/view/PXB%208.0/job/percona-xtrabackup-8.0-multijob/326/

satya-bodapati commented on storage/innobase/xtrabackup/test/t/pxb-2797_import_fulltext_schema_mismatch.sh:
```diff
@@ -0,0 +1,54 @@
+mysql test <<EOF
+CREATE TABLE test.t1 (
+  c1 int NOT NULL AUTO_INCREMENT,
+  c2 varchar(50) DEFAULT NULL,
```
better remove default null and insert some valid values.

satya-bodapati commented on storage/innobase/xtrabackup/test/t/pxb-2797_import_fulltext_schema_mismatch.sh:
```diff
@@ -0,0 +1,54 @@
+
+CREATE TABLE test.t2 (
+  c1 int NOT NULL AUTO_INCREMENT,
+  c2 bigint DEFAULT NULL,
+  c3 bigint DEFAULT NULL,
```
same here better remove default null values and insert some valid values

satya-bodapati commented on storage/innobase/xtrabackup/test/t/pxb-2797_import_fulltext_schema_mismatch.sh:
```diff
@@ -0,0 +1,54 @@
+  PRIMARY KEY (c1),
+  INDEX idx1 (c2),
+  INDEX idx2 (c2 DESC),
+  INDEX idx3 (c2, c3 DESC)
```
please use c2 ASC, c3 DESC

satya-bodapati replied:
also make one index as unique index

satya-bodapati commented on storage/innobase/xtrabackup/test/t/pxb-2797_import_fulltext_schema_mismatch.sh:
```diff
@@ -0,0 +1,54 @@
+  INDEX idx2 (c2 DESC),
+  INDEX idx3 (c2, c3 DESC)
+) ENGINE=InnoDB;
+INSERT INTO test.t2 (c2) values (1000);
+INSERT INTO test.t2 (c2) values (2000);
+INSERT INTO test.t2 (c2) values (3000);
+INSERT INTO test.t2 (c3) values (4000);
+INSERT INTO test.t2 (c3) values (5000);
+INSERT INTO test.t2 (c3) values (6000);
```
can you please use valid values for indexes

aybek commented:
https://pxb.cd.percona.com/job/percona-xtrabackup-8.0-multijob/328/

satya-bodapati commented on storage/innobase/dict/dict0dd.cc:
```diff
@@ -981,7 +982,11 @@ dict_table_t *dd_table_create_on_dd_obj(const dd::Table *dd_table,
       }
 
-      bool is_asc = (idx_elem->order() == dd::Index_element::ORDER_ASC);
+      /* FULLTEXT and HASH indexes can have UNDEF order, we should treat UNDEF
```
Just realized that you mentioned HASH indexes here , can you please cover this in the testcase?
Also add a spatial index as well https://dev.mysql.com/doc/refman/8.0/en/creating-spatial-indexes.html

satya-bodapati reviewed (changes requested):
Almost there. I just rememberd one more scenario that the testcase should cover.

satya-bodapati commented on storage/innobase/xtrabackup/test/t/xb_export.sh:
```diff
@@ -244,3 +244,70 @@ vlog "Checksums are OK"
+cp $topdir/backup/test/t3.cfg $mysql_datadir/test/t3.cfg
+mysql -e "ALTER TABLE test.t3 IMPORT TABLESPACE;"
+
+verify_db_state test
```
please add rm $topdir/backup here

satya-bodapati reviewed (approved):
LGTM! (one minor comment in testcase)

aybek committed:
PXB-2797: Schema mismatch when importing table with full-text index from xtrabackup backup

https://jira.percona.com/browse/PXB-2797

When importing a single table (IMPORT TABLESPACE) from a backup made using xtrabackup and the table contains a full-text index the import process will error out with:
`ERROR 1808 (HY000) at line 132: Schema mismatch (Index xxxxxx field xxxxxx is ascending which does not match metadata file which is descending)`

The problem occurs because FTS tables are created internally by InnoDB Storage engine and they do not choose any ordering flag. Default is UNDEF.
This patch solves the issue by treating UNDEF order as ASC when creating dict_table_t on `--prepare` step.
